### ISSUE TITLE
[BUG] Engine stuck in "stopped" state, prevent volume attach

### ISSUE THREAD

**Issue by tmmorin** [longhorn/longhorn#9938]
## Describe the bug

We observe cases where one or more Engines (and corresponding Replicas) are stuck in "stopped" state, and the corresponding Volumes can't be attached to pods.

## To Reproduce

This issue has frequently been observed in our project ([Sylva](https://gitlab.com/sylva-projects/sylva-core)) in nightly CI pipelines since we upgraded from 1.6.3 to 1.7.2.

The issue happens during phase where all Nodes are incrementally replaced (we use Cluster API which updates the nodes in rolling updates).

We don't have a recipe to reproduce but [the issue we filed on our side](https://gitlab.com/sylva-projects/sylva-core/-/issues/1921) has more details and quite extensive dumps/logs.

## Expected behavior

We expect Longhorn to remain fully available during Cluster API Node rolling updates.

## Support bundle for troubleshooting

The closest thing we have is our CI job artifacts which have extensive resource dumps, pod logs, and events.
They can be accessed via the "download" button on the following URLs:
* https://gitlab.com/sylva-projects/sylva-core/-/jobs/8559500297
* https://gitlab.com/sylva-projects/sylva-core/-/jobs/8578325612

## Environment

 - Longhorn version: 1.7.2
 - Impacted volume (PV): depends on occurence, but for instance pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-e-0 on https://gitlab.com/sylva-projects/sylva-core/-/jobs/8559500297
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher chart
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2 1.30.5 or 1.30.6
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 1
 - Node config
   - OS type and version: openSUSE Leap 15.6
   - Kernel version: 6.4.0-150600.23.25-default
   - CPU per node:
   - Memory per node:
   - Disk type (e.g. SSD/NVMe/HDD):
   - Network bandwidth between the nodes (Gbps):
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal):   KVM 
 - Number of Longhorn volumes in the cluster: 38

## Workaround and Mitigation

None known

**Comment by derekbit** [comment]
cc @ChanYiLin

**Comment by derekbit** [comment]
Some findings
- Setting `allow-volume-creation-with-degraded-availability` is `false`
- The `volume.status.condition["Scheduled"]` of `pvc-9d02c710-a934-40e1-88aa-59997fadc1e9` is `false`, because the number of replicas (=1) and volume.spec.numOfReplicas (=2) is not equal. ([codes](https://github.com/longhorn/longhorn-manager/blob/v1.7.2/controller/volume_controller.go#L1769-L1771))
- The conditions `volume.status.currentState = detached` and `volume.status.condition["Scheduled"] = false` results in the failure of the volume detachment ([codes](https://github.com/longhorn/longhorn-manager/blob/v1.7.2/api/model.go#L1526)) and emits the error message `"volume ...  is not ready for workloads"` ([codes](https://github.com/longhorn/longhorn-manager/blob/v1.7.2/csi/controller_server.go#L461-L463)).

@tmmorin 
- What's the purpose of adjusting the setting `allow-volume-creation-with-degraded-availability` is `false`? By default, the value is `true`.
- Could you change the setting `allow-volume-creation-with-degraded-availability` to `true` for checking the issue can be resolved?

**Comment by tmmorin** [comment]
About allow-volume-creation-with-degraded-availability: I think we just follow documented best practices, which say:

> It’s recommended to disable this setting when using Longhorn in the production environment. See Best Practices for details.

(https://longhorn.io/docs/1.7.2/references/settings/#allow-volume-creation-with-degraded-availability) 

But yes, we can try to disable this setting. 

I can understand why this setting would make convergence slower during a node rolling update, but I don't get why the platform would get stuck. 

Would you have a scenario in mind that would explain, with this setup, why the platform does not eventually recover?

**Comment by ChanYiLin** [comment]
Hi @tmmorin
We are also reviewing if this recommendation is appropriate.

Would like to know the details of how "the platform" got stuck.
When this setting is set to "false", that means if the volume don't have enough healthy replicas, it won't be able to attach and mount to the pod.
That was why the Engine got stuck in "Stopped" state.
So I guess the "the platform" getting stuck makes sense?

**Comment by tmmorin** [comment]
> Would like to know the details of how "the platform" got stuck. When this setting is set to "false", that means if the volume don't have enough healthy replicas, it won't be able to attach and mount to the pod. That was why the Engine got stuck in "Stopped" state. So I guess the "the platform" getting stuck makes sense?

I would expect Longhorn to have, after some time, rebuilt enough healthy replicas which would have allowed the platform to recover.  

But on hour runs this "never" happens ; here "never" means that the CI job times out ...

in https://gitlab.com/sylva-projects/sylva-core/-/jobs/8592499658 the `FailedAttachVolume` start at 2024-12-09T22:59:38 and the job times out at 2024-12-10T00:50:07  ... so this means nearly 2 hours with the Engine remaining in `stopped` state.

**Comment by tmmorin** [comment]
Also worth noting: we've been using `allow-volume-creation-with-degraded-availability: false` for a long time (nearly 1 year) and we've been having able to to Cluster API node rolling updates  with both Longhorn 1.5 and 1.6.

**Comment by derekbit** [comment]
> Also worth noting: we've been using `allow-volume-creation-with-degraded-availability: false` for a long time (nearly 1 year) and we've been having able to to Cluster API node rolling updates with both Longhorn 1.5 and 1.6.

@tmmorin 
Thanks for the valuable information. This is a regression after introducing the [PR](https://github.com/longhorn/longhorn-manager/commit/877e7cba36dcc238c6744f8958998307e87ad1ce) in v1.7.2.

I can reproduce the issue by the following steps
1. Create a 2-node Longhorn cluster
2. Create a workload pod with a 2-replica volume 
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: longhorn-test
provisioner: driver.longhorn.io
allowVolumeExpansion: true
reclaimPolicy: Delete
volumeBindingMode: Immediate
parameters:
  numberOfReplicas: "2"
  staleReplicaTimeout: "2880"
  fromBackup: ""
  fsType: "ext4"
---
---
apiVersion: v1
kind: Pod
metadata:
  name: volume-test
  namespace: default
spec:
  restartPolicy: Always
  containers:
  - name: volume-test
    image: nginx:stable-alpine
    imagePullPolicy: IfNotPresent
    livenessProbe:
      exec:
        command:
          - ls
          - /data/lost+found
      initialDelaySeconds: 5
      periodSeconds: 5
    volumeMounts:
    - name: volv
      mountPath: /data
    ports:
    - containerPort: 80
  volumes:
  - name: volv
    persistentVolumeClaim:
      claimName: longhorn-volv-pvc
```
3. The volume is attached and then update `volume.spec.numberOfReplicas` to 3
4. The number of replicas is still 2 because there are only two nodes. (checked by `kubectl -n longhorn-system get replicas`)
5. Because of https://github.com/longhorn/longhorn-manager/blob/v1.7.2/controller/volume_controller.go#L1769-L1771, the `volume.status.condition["Scheduled"]` is `false`. (checked by `kubectl -n longhorn-system get volumes <name>`)
6. Detach the volume.
7. Try to attach the volume, the attachment should fail because of https://github.com/longhorn/longhorn-manager/blob/v1.7.2/api/model.go#L1526.

We will fix the issue in the upcoming v1.8.0 and v1.7.3.

As @ChanYiLin and I mentioned, the current workaround is enabling `allow-volume-creation-with-degraded-availability`.  The volume controller will set the `Scheduled` condition to `true` and prevent the attachment to be stuck forever.


cc @c3y1huang

**Comment by tmmorin** [comment]
Many thanks for this update @derekbit  !

We would like to volunteer testing the fix as soon as possible. Are there container image nightly builds that we could use (once the fix lands) ?

**Comment by tmmorin** [comment]
Another question: do you think this issue could also cause #9939 ?

**Comment by derekbit** [comment]
> Many thanks for this update @derekbit !
> 
> We would like to volunteer testing the fix as soon as possible. Are there container image nightly builds that we could use (once the fix lands) ?

Sure! We will fix the issue and provide a custom image for testing. Many thanks.

 > Another question: do you think this issue could also cause https://github.com/longhorn/longhorn/issues/9939 ?
 
We will investigate the issue soon.
cc @PhanLe1010 Could you help check https://github.com/longhorn/longhorn/issues/9939? Thank you.

**Comment by ChanYiLin** [comment]
I think the current situation can be simplify to this

- If you have 1 node
- You create 1 volume with 2 replicas
- So there is 1 replica can't be scheduled
- You try to attach the volume, it will fail because of the setting `allow-volume-creation-with-degraded-availability = false`
- The engine will get stuck in stopped state
- Then, even if you add another node to the cluster, Longhorn won't replenish another replica because the volume is not in attached state, and it won't be able to do the online rebuilding to rebuild new replica.

That is why in your situation, it stuck forever.

Can you help us to turn on the setting to see if the issue is resolved.

For "Why it didn't happen before?", can you provide any successful CI run in Longhorn version 1.6, so we can compare the behavior?
We didn't change the behavior of this setting since it was there. So we may need more information.
Thanks

**Comment by tmmorin** [comment]
> Can you help us to turn on the setting to see if the issue is resolved.

I have this in flight: https://gitlab.com/sylva-projects/sylva-core/-/merge_requests/3473 ("pipelines" tab) where I started 12 runs. We'll see how it goes.

> For "Why it didn't happen before?", can you provide any successful CI run in Longhorn version 1.6, so we can compare the behavior?

Sure, here is for instance one from last night: https://gitlab.com/sylva-projects/sylva-core/-/jobs/8604708263

**Comment by ChanYiLin** [comment]
@tmmorin 
Thanks for the information!

**Comment by innobead** [comment]
Let's add the test case described in https://github.com/longhorn/longhorn/issues/9938#issuecomment-2535506472 by @derekbit . 

cc @longhorn/qa

**Comment by tmmorin** [comment]
We have been trying 1.7.1 to see if the problem would occur or not.

And it seems that the problem *does* occur with 1.7.1

In run https://gitlab.com/sylva-projects/sylva-core/-/jobs/ you we see that in metal3-system namespace we have pods that can't attach a longhorm volume:

```
2024-12-10T13:22:22Z	2024-12-10T13:53:00Z	attachdetach-controller-	Pod	metal3-metal3-ironic-8448cbb4f-5frfr	16	FailedAttachVolume	"AttachVolume.Attach failed for volume ""pvc-45c7a1ba-f52d-42d6-aab8-f729c891a461"" : rpc error: code = DeadlineExceeded desc = volume pvc-45c7a1ba-f52d-42d6-aab8-f729c891a461 failed to attach to node mgmt-1581894285-rke2-capm3-virt-management-md-0 with attachmentID csi-5898653a743a5acad3d75a76962edbc985a744bee7085960ca0bf45b34e11c10"
2024-12-10T13:22:22Z	2024-12-10T13:53:01Z	attachdetach-controller-	Pod	metal3-metal3-mariadb-647cf958c4-8z85r	16	FailedAttachVolume	"AttachVolume.Attach failed for volume ""pvc-6f6f5692-79a2-4df0-969b-c40fa5e83e90"" : rpc error: code = DeadlineExceeded desc = volume pvc-6f6f5692-79a2-4df0-969b-c40fa5e83e90 failed to attach to node mgmt-1581894285-rke2-capm3-virt-management-md-0 with attachmentID csi-e62a1c8cd468558614c96e8eaa629bcd5fd205955c9866313eca22fd561a1254"
```

and `Engines.longhorn.io.summary.txt` shows that 11 Engines are in `stopped` state

So this seems to contradict  the hypothesis that you made above, @derekbit, that https://github.com/longhorn/longhorn-manager/commit/877e7cba36dcc238c6744f8958998307e87ad1ce,  introduced in 1.7.2, is the change causing the issue.

/cc @zioc

**Comment by derekbit** [comment]
> @derekbit
> 
> > As @ChanYiLin and I mentioned, the current workaround is enabling allow-volume-creation-with-degraded-availability.
> > @ChanYiLin
> > Can you help us to turn on the setting to see if the issue is resolved.
> 
> I was looking at the last batch of runs finishing, and _nearly_ going to post a "this works!" message.
> 
> **BUT** 😭, in a total of 20 jobs, **one failed**, with all the symptoms of this issue:
> 
> * `metal3-metal3-mariadb-5f65f55f88-949vj` pod in `metal3-system` failing, with this event
> 
> ```
> 2024-12-12T10:42:44Z	2024-12-12T12:12:37Z	attachdetach-controller-	Pod	metal3-metal3-mariadb-5f65f55f88-949vj	42	FailedAttachVolume	"AttachVolume.Attach failed for volume ""pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9"" : rpc error: code = DeadlineExceeded desc = volume pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9 failed to attach to node mgmt-1585382134-rke2-capm3-virt-management-md-0 with attachmentID csi-4ac237ef63c0b44e06131d00aaddc109e37db6860c0ae27ef49c48ff50e8aaef"
> ```
> 
> * 11 Engines in `stopped` state
> 
> Detailed dump are in [gitlab.com/sylva-projects/sylva-core/-/jobs/8621059119](https://gitlab.com/sylva-projects/sylva-core/-/jobs/8621059119) (under `management-cluster-dump`)

Thanks for the support bundle and I will investigate it. This should be caused by another issue because the error message is different than the previous case. 
BTW, the failure rate decreases in v1.7.1, correct?

**Comment by tmmorin** [comment]
Thanks for your reply.

>  This should be caused by another issue because the error message is different than the previous case.

Ok.

>  BTW, the failure rate decreases in v1.7.1, correct?

I have the same feeling. But maybe we don't have enough data points.

**Comment by ChanYiLin** [comment]
After discussed with @derekbit 
For the volume becomes `NotReadyForWorkload` and not able to attach due to `Scheduled=False` in the condition

To summarize
1. Longhorn creates the replicas when the volume is created
2. All the replicas should be scheduled when created, that is, the `NodeID` in the replica.Spec should be assigned.
3. We iterate all the replicas, if there is any replica not scheduled
    - We set the `Scheduled=False`
    - The volume is  `NotReadyForWorkload` and cannot be used
    - However, if set the `Allow Volume Creation with Degraded Availability` to `True`, then the volume is still allowed to use.

Until here, it is all the same for all the versions
Now, in users case, the node goes down due to upgrade, so the replica on the node becomes failed

And we have a logic which will delete the replica CR directly it it is in failed state for a while
Thus the replica  `pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-r-f56f974e` is deleted
```
time="2024-12-05T18:23:30Z" level=warning msg="Replica pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-r-f56f974e failed too long ago to be useful during a rebuild" func="controller.(*VolumeController).shouldCleanUpFailedReplica" file="volume_controller.go:4884" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=mgmt-1575464159-rke2-capm3-virt-management-cp-0 owner=mgmt-1575464159-rke2-capm3-virt-management-cp-0 replica=pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-r-f56f974e state=detached volume=pvc-9d02c710-a934-40e1-88aa-59997fadc1e9
time="2024-12-05T18:23:30Z" level=info msg="Cleaning up corrupted, staled replica" func="controller.(*VolumeController).cleanupCorruptedOrStaleReplicas" file="volume_controller.go:1022" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=mgmt-1575464159-rke2-capm3-virt-management-cp-0 owner=mgmt-1575464159-rke2-capm3-virt-management-cp-0 replica=pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-r-f56f974e state=detached volume=pvc-9d02c710-a934-40e1-88aa-59997fadc1e9
```

### Before v1.7.2, before the [PR](https://github.com/longhorn/longhorn-manager/commit/877e7cba36dcc238c6744f8958998307e87ad1ce)

1. We only iterate the replica CRs and see if all of them are scheduled
2. And yes, the failed one is deleted and the replicas left are still scheduled, so the `Scheduled=` is still True
3. So the volume can still be attached after awhile, and replenish and rebuild the replica to match the spec.numberOfReplicas\
4. Everything works fine.

### After the PR

1. We now not only iterate the existing CRs, but we also compare the `len(rs) == spec.numberOfReplicas` to decide if `Scheduled` is True or False
2. And no, **the failed replica CR was deleted and was not replenished because the volume was detached.** So the number did not match
3. `Scheduled` was False and the volume was `NotReadyForWorkload`, so the volume coulnd't attach to the workload and stuck forever.

cc @derekbit @c3y1huang @tmmorin 


---
Hi @c3y1huang , We can discuss if we really need to compare the `len(rs)` to decide if `Scheduled` is True or False. There could have chance that the replica CR is deleted due to some failure handling.

**Comment by derekbit** [comment]
@tmmorin 
We've scheduled v1.7.3. release on February 20, 2025. Before the release, we will check the feasibility of the solution, if it works and is without any side effect, we will build a custom image for your test.

cc @ChanYiLin @c3y1huang @innobead

**Comment by innobead** [comment]
Thanks for the cooperation @tmmorin

**Comment by c3y1huang** [comment]
> Hi @c3y1huang , We can discuss if we really need to compare the `len(rs)` to decide if `Scheduled` is True or False. There could have chance that the replica CR is deleted due to some failure handling.
> 
> After some studying, we now understand that.
> 
> 1. We now have a precheck before creating replica which will also consider the scheduling
> 2. So we won't create the replica if there is no enough disk
> 3. So we will need extra len(rs) check to decide if `Scheduled` is True or False
>    May need to find another way to refine the flow.

We need to include the `len(rs)` condition check for handling the volume condition when replicas are not created due to the precheck. After discussing with @derekbit , we believe the issue can be addressed by enhancing [this section](https://github.com/longhorn/longhorn-manager/blob/v1.8.0-dev-20241215/controller/volume_controller.go#L1759-L1779) to account for existing volumes.

**Comment by derekbit** [comment]
> @derekbit
> 
> > As @ChanYiLin and I mentioned, the current workaround is enabling allow-volume-creation-with-degraded-availability.
> > @ChanYiLin
> > Can you help us to turn on the setting to see if the issue is resolved.
> 
> I was looking at the last batch of runs finishing, and _nearly_ going to post a "this works!" message.
> 
> **BUT** 😭, in a total of 20 jobs, **one failed**, with all the symptoms of this issue:
> 
> * `metal3-metal3-mariadb-5f65f55f88-949vj` pod in `metal3-system` failing, with this event
> 
> ```
> 2024-12-12T10:42:44Z	2024-12-12T12:12:37Z	attachdetach-controller-	Pod	metal3-metal3-mariadb-5f65f55f88-949vj	42	FailedAttachVolume	"AttachVolume.Attach failed for volume ""pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9"" : rpc error: code = DeadlineExceeded desc = volume pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9 failed to attach to node mgmt-1585382134-rke2-capm3-virt-management-md-0 with attachmentID csi-4ac237ef63c0b44e06131d00aaddc109e37db6860c0ae27ef49c48ff50e8aaef"
> ```
> 
> * 11 Engines in `stopped` state
> 
> Detailed dump are in [gitlab.com/sylva-projects/sylva-core/-/jobs/8621059119](https://gitlab.com/sylva-projects/sylva-core/-/jobs/8621059119) (under `management-cluster-dump`)

### RCA
- Problematic volume: pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9
- Error message
	```
	time="2024-12-12T12:13:59Z" level=warning msg="Replica pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9-r-7fad0729 is marked as failed because the volume pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9 is not attached and the instance manager is unable to launch the replica" ....
	``` 

- Replica `pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9-r-7fad0729` is on node `mgmt-1585382134-rke2-capm3-virt-management-cp-2`, but the instance-manager on this node is in `error` state. The `error` state is due to the node `mgmt-1585382134-rke2-capm3-virt-management-cp-2` is cordoned, so the corresponding pod is unable to start.

```
    name: instance-manager-105930e30b903199fd48165e1b56e78a
    namespace: longhorn-system
    ownerReferences:
    - apiVersion: longhorn.io/v1beta2
      blockOwnerDeletion: true
      kind: Node
      name: mgmt-1585382134-rke2-capm3-virt-management-cp-2
      uid: 3608a8f1-4d23-4237-afb2-d878af533393
    resourceVersion: "420802"
    uid: 6e7b9b56-d1dc-4cdd-8c90-d46ff52d89fd
  spec:
    dataEngine: v1
    image: rancher/mirrored-longhornio-longhorn-instance-manager:v1.7.2
    nodeID: mgmt-1585382134-rke2-capm3-virt-management-cp-2
    type: aio
  status:
    apiMinVersion: 0
    apiVersion: 0
    currentState: error
    ip: 100.72.12.24
    ownerID: mgmt-1585382134-rke2-capm3-virt-management-cp-2
    proxyApiMinVersion: 1
    proxyApiVersion: 5
```

@tmmorin Should the node `mgmt-1585382134-rke2-capm3-virt-management-cp-2` be cordoned?

**Comment by tmmorin** [comment]
hello

@derekbit:
> Should the node mgmt-1585382134-rke2-capm3-virt-management-cp-2 be cordoned?

Yes, this is expected, this is a node drain+cordon operation (one step of the Cluster API node rolling update being done, right before tearing down a node).

**Comment by derekbit** [comment]
> Yes, this is expected, this is a node drain+cordon operation (one step of the Cluster API node rolling update being done, right before tearing down a node).

@tmmorin
Thanks for the update.
The volume `pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9` should have 3 replicas, but the current status is it has only one replica and it is on the cordoned node. Before cordoning the node, it should make sure other replicas of the volume are rebuilt on the nodes.

**Comment by tmmorin** [comment]
> it should make sure other replicas of the volume are rebuilt on the nodes.

Just to be sure we're aligned: so the anomaly is that Longhorn should have rebuilt other Replicas but did not ?  (or do you imply that there is something that our system does not ensure and should be ensuring ?)

**Comment by derekbit** [comment]
> Just to be sure we're aligned: so the anomaly is that Longhorn should have rebuilt other Replicas but did not ? (or do you imply that there is something that our system does not ensure and should be ensuring ?)

No, Longhorn will rebuild replicas if they are lost. However, the rebuilding process may be delayed for various reasons, such as the inability to find a suitable node or disk temporarily.

To better understand your steps, can you elaborate more on the steps that were applied the node `mgmt-1585382134-rke2-capm3-virt-management-cp-2` and which step it is stuck in? Thankyou.

**Comment by tmmorin** [comment]
> No, Longhorn will rebuild replicas if they are lost. However, the rebuilding process may be delayed for various reasons, such as the inability to find a suitable node or disk temporarily.

Ok.

But in that run the pod trying to mount that PVC had the failure event firing for a long time: 42 times, between 2024-12-12T10:42:44Z and 2024-12-12T12:12:37Z (90 minutes).

> To better understand your steps and find out the root cause, can you elaborate more on the steps that were applied the node mgmt-1585382134-rke2-capm3-virt-management-cp-2 and which step it is stuck in? Thank you.

The platform was stuck in the following state: following the drain of mgmt-1585382134-rke2-capm3-virt-management-cp-2, cluster API main controller was trying to delete this machines; for this it relies on the metal3 component, but this component was done because the pod of its database backend (metal3-metal3-mariadb-5f65f55f88-949vj) could not start, which was because pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9 could not be attached (this PVC had its Engine in 'stopped' state).

Before that, two nodes of the platform had been successfully subject to the rolling update (mgmt-1585382134-rke2-capm3-virt-management-cp-1 and mgmt-1585382134-rke2-capm3-virt-management-md-0).

**Comment by derekbit** [comment]
> > No, Longhorn will rebuild replicas if they are lost. However, the rebuilding process may be delayed for various reasons, such as the inability to find a suitable node or disk temporarily.
> 
> Ok.
> 
> But in that run the pod trying to mount that PVC had the failure event firing for a long time: 42 times, between 2024-12-12T10:42:44Z and 2024-12-12T12:12:37Z (90 minutes).
> 
> > To better understand your steps and find out the root cause, can you elaborate more on the steps that were applied the node mgmt-1585382134-rke2-capm3-virt-management-cp-2 and which step it is stuck in? Thank you.
> 
> The platform was stuck in the following state: following the drain of mgmt-1585382134-rke2-capm3-virt-management-cp-2, cluster API main controller was trying to delete this machines; for this it relies on the metal3 component, but this component was done because the pod of its database backend (metal3-metal3-mariadb-5f65f55f88-949vj) could not start, which was because pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9 could not be attached (this PVC had its Engine in 'stopped' state).
> 
> Before that, two nodes of the platform had been successfully subject to the rolling update (mgmt-1585382134-rke2-capm3-virt-management-cp-1 and mgmt-1585382134-rke2-capm3-virt-management-md-0).

One more question for checking why the replicas of the volume are all in failed state.

`pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9` has three replicas
- `pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9-r-7fad0729` on `mgmt-1585382134-rke2-capm3-virt-management-cp-2`
- `pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9-r-a8001515` on `mgmt-1585382134-rke2-capm3-virt-management-cp-1`
- `pvc-88341e3d-bd4c-4fa2-9fae-4043373b71e9-r-df2f1974` on `mgmt-1585382134-rke2-capm3-virt-management-cp-0`

- While draining the nodes, was the volume attached? 
- How did you cordon and drain the node? Can you provide the command you executed?

**Comment by tmmorin** [comment]
> While draining the nodes, was the volume attached?

It was attached before the drain (the metal3-metal3-ironic-xxxx pod needs it and I know it was up or else the node rolling update can't proceed). 

>  How did you cordon and drain the node? Can you provide the command you executed?

We don't execute a command.

This is done automatically by Cluster API Machine controller.

It's very similar to how the ReplicaSet controller works: a Machine template is changed, a Machine is cordoned/drained, then deleted, and then  a new one is created from the new template... all machines are replaced like this until all are based on the new template.

The details of how cordon/drain is done are in Cluster API code, but really this is only very standard things :  node taints + eviction API.

**Comment by derekbit** [comment]
> It was attached before the drain (the metal3-metal3-ironic-xxxx pod needs it and I know it was up or else the node rolling update can't proceed).

IIUC, the voluem shouldn't be detached during rolling update? Am I right?

**Comment by tmmorin** [comment]
## Status update on the result of the testing we did with Longhorn 1.7.2 with allowVolumeCreationWithDegradedAvailability=true 

We ran a fairly large batch of CI pipelines during the last week-end (more than 70). And in this batch none seemed to expose the problem discussed here in this issue (nor the one in #9939).

(we noticed one Longhorn instance-manager pod failing to be evicted, we'll file a separate issue on this topic)

So from what we can see in our CI, using allowVolumeCreationWithDegradedAvailability=true is an acceptable workaroun; we'll rely on this until a fix is implemented, hopefully for 1.7.3 release.

**Comment by ChanYiLin** [comment]
Hi @tmmorin 
From the support bundle, we found there are lots of volumes had no healthy replicas at some point
This makes Longhorn very unstable and increases the likelihood of data loss.

When you do rolling upgrade to the cluster
Do you wait until the node is fully drained, with no replicas or engines running on it, before deleting the node?
Please refer to our [Node Maintenance and Kubernetes Upgrade Guide
](https://longhorn.io/docs/1.7.2/maintenance/maintenance/#updating-the-node-os-or-container-runtime) to ensure your rolling upgrade process aligns with the instructions provided.

Second question, according to the docs.
> Note: By default, if there is one last healthy replica for a volume on the node, Longhorn will prevent the node from completing the drain operation, to protect the last replica and prevent the disruption of the workload. You can control this behavior with the setting [Node Drain Policy](https://longhorn.io/docs/1.7.2/references/settings#node-drain-policy), or [evict the replica to other nodes before draining](https://longhorn.io/docs/1.7.2/nodes-and-volumes/nodes/disks-or-nodes-eviction). See [Node Drain Policy Recommendations](https://longhorn.io/docs/1.7.2/maintenance/maintenance/#node-drain-policy-recommendations) for considerations when selecting a policy.
Could you please check if the setting is exactly what you need?

**Comment by derekbit** [comment]
Hello @tmmorin Any update?

**Comment by ChanYiLin** [comment]
Hi @tmmorin 
Could you help us test the image `jackfantasy/longhorn-manager:9937_fix_engine_stop` to see if it resolves the issue, even when the setting `Allow Volume Creation with Degraded Availability` is set to False?
To do so, please use `kubectl edit daemonset/longhorn-manager -n longhorn-system`.
Note that two places need to be updated with the new image.
1. `.spec.template.spec.containers[0].command["--manager-image"]`
2. `.spec.template.spec.containers[0].image`

Thanks

**Comment by ChanYiLin** [comment]
Hi @tmmorin 
Two more questions
1. Is `monitoring-pool-0-1` this pod bound to the node `mgmt-1575464159-rke2-capm3-virt-management-md-0`? I saw there are affinity settings. Does that means during the rolling upgrade, the pod won't start before the node `...-md-0` is backup?
2. I found there are two nodes being drained at the same time: `...-md-0` and `...-cp-1`, shouldn't it be rolling upgrade one by one?
```
I1205 17:52:42.710684       1 machine_controller.go:388] "Draining node" controller="machine" controllerGroup="cluster.x-k8s.io" controllerKind="Machine" Machine="sylva-system/mgmt-1575464159-rke2-capm3-virt-md0-slfdj-vqqb9" namespace="sylva-system" name="mgmt-1575464159-rke2-capm3-virt-md0-slfdj-vqqb9" reconcileID="c9a4040d-d60e-4abd-9093-55a3533237a1" MachineSet="sylva-system/mgmt-1575464159-rke2-capm3-virt-md0-slfdj" MachineDeployment="sylva-system/mgmt-1575464159-rke2-capm3-virt-md0" Cluster="sylva-system/mgmt-1575464159-rke2-capm3-virt" Node="mgmt-1575464159-rke2-capm3-virt-management-md-0"

I1205 17:52:43.367613       1 machine_controller.go:388] "Draining node" controller="machine" controllerGroup="cluster.x-k8s.io" controllerKind="Machine" Machine="sylva-system/mgmt-1575464159-rke2-capm3-virt-control-plane-dmjg8" namespace="sylva-system" name="mgmt-1575464159-rke2-capm3-virt-control-plane-dmjg8" reconcileID="8d8a4fcf-c119-44a1-81a3-caa67b510d2e" RKE2ControlPlane="sylva-system/mgmt-1575464159-rke2-capm3-virt-control-plane" Cluster="sylva-system/mgmt-1575464159-rke2-capm3-virt" Node="mgmt-1575464159-rke2-capm3-virt-management-cp-1"
```

**Comment by ChanYiLin** [comment]
Hi @PhanLe1010 
Would like to invite your to take a look at this issue. It is also related to the replica scheduling.
However there are still something I couldn't figure out.

The support bundle I was studying was 
> Impacted volume (PV): depends on occurence, but for instance pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-e-0 on https://gitlab.com/sylva-projects/sylva-core/-/jobs/8559500297

Here is my note
[note.txt](https://github.com/user-attachments/files/18515342/note.txt)

From the support bundle I could see that two nodes were cordoned and drained at the same time when rolling upgrade the cluster.
The effected volume was attached at the same time when the node was cordoned
Since it had best-effort locality, there was a replica `pvc-9d02c710-a934-40e1-88aa-59997fadc1e9-r-e9a08274` created to the cordoned node when the volume was attached. Thus, the volume got the condition `LocalReplicaSchedulingFailure` due to the replica couldn't get scheduled to the affinity node.

After a while, when the node was drained, the volume was detached again.
The mystery was that two replicas CR including the unscheduled one were gone without any deletion logs. Only one replica left.
However, since the `len(rs) < numberOfReplicas`, the volume condition couldn't not reset to `scheduled` and thus remained  `notReadyForWorkload` when csi tried to attach the volume.
The volume couldn't be used anymore unless user turned on the setting `Allow Volume Creation with Degraded Availability`

I have more detailed notes in the note.txt, PTAL, thanks

**Comment by PhanLe1010** [comment]
Surely @ChanYiLin Checking this

**Comment by PhanLe1010** [comment]
> The effected volume was attached at the same time when the node was cordoned

Hi @ChanYiLin 

Quick question, how do you know that the volume is trying to attach to the node `md-0` at the same time?

**Comment by PhanLe1010** [comment]
> The mystery was that two replicas CR including the unscheduled one were gone without any deletion logs. Only one replica left.

Looks like the longhorn-manager pod on the node `mgmt-1575464159-rke2-capm3-virt-management-md-0` is recently recreated so you cannot find the logs. Its first logs is after the incident at `2024-12-05T18:00:20Z`
`time="2024-12-05T18:00:20Z" level=info msg="Starting longhorn conversion webhook server" func=webhook.StartWebhook file="webhook.go:24"`

**Comment by PhanLe1010** [comment]
## Found reproducing method:

1. Create a cluster of 3 nodes and set `Allow Volume Creation with Degraded Availability` to `false`
1. Create a volume with:
    1. Data locality: best-effort
    1. Number of replicas: 2
    1. Note that if you create the volume using Longhron UI, it seems to set the `StaleReplicaTimeout` to 20 mins. Also note that this value in @tmmorin 's cluster is 30 mins
1. Assume that `replica-1` is on `node-1`, `replica-2` is on `node-2`
1. Cordon `node-3` then attach the volume to `node-3`
1. Observe that Longhorn tries to create a `replica-3` with `HardNodeAffinity` set to `node-3`. However, `node-3` is currently unschedulable because it was cordoned. Longhorn volume has `ReplicaSchedulingFailure` condition like this:
    ```yaml
        - lastProbeTime: ''
          lastTransitionTime: '2025-01-28T07:06:39Z'
          message: hard affinity cannot be satisfied
          reason: LocalReplicaSchedulingFailure
          status: 'False'
    ```
1. Cordon `node-2` and delete the instance manager pod on `node-2`. 
1. Observe that note that `replica-2` becomes failed. However, Longhorn does not rebuild the `replica-2` because `node-2` was cordoned 
1. Wait 21 mins for `replica-2` to become stale and removed by Longhorn. Observe this message (similar to the logs in the support bundle provided by @tmmorin at https://gitlab.com/sylva-projects/sylva-core/-/jobs/8559500297):
    `
    [longhorn-manager-m2k9n longhorn-manager] time="2025-01-28T07:43:50Z" level=warning msg="Replica testvol-r-07602639 failed too long ago to be useful during a rebuild" func="controller.(*VolumeController).shouldCleanUpFailedReplica" file="volume_controller.go:4890" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=phan-v713-pool2-5nr9q-78qnr owner=phan-v713-pool2-5nr9q-78qnr replica=testvol-r-07602639 state=attached volume=testvol
    `
    Ref the code: https://github.com/longhorn/longhorn-manager/blob/fd2ef9c5ee891fdbc787e475359a48b3f6b0e094/controller/volume_controller.go#L4882-L4886
1. Detach the volume
1. Observer that Longhorn delete `replica-3` (the unschedulable replica with with `HardNodeAffinity` set to `node-3`):
    `
    [longhorn-manager-m2k9n longhorn-manager] time="2025-01-28T07:43:50Z" level=warning msg="Replica testvol-r-07602639 failed too long ago to be useful during a rebuild" func="controller.(*VolumeController).shouldCleanUpFailedReplica" file="volume_controller.go:4890" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=phan-v713-pool2-5nr9q-78qnr owner=phan-v713-pool2-5nr9q-78qnr replica=testvol-r-07602639 state=attached volume=testvol
    `
    See more details about why Longhorn delete this replica in the additional context below
1. The volume now has exact same condition as in the user's support bundle:
    ```yaml
        - lastProbeTime: ''
          lastTransitionTime: '2025-01-28T07:06:39Z'
          message: replica scheduling failed
          reason: LocalReplicaSchedulingFailure
          status: 'False'
          type: Scheduled
    ```
1. From now on, workload cannot attach the volume anymore because the volume is stuck in this `LocalReplicaSchedulingFailure` condition forever. 

### Additional Context
1. When Longhorn create a new replica and if the volume is not a newly created volume, Longhorn set `replica.Spec.RebuildRetryCount = scheduler.FailedReplicaMaxRetryCount` which is 5. Ref code https://github.com/longhorn/longhorn-manager/blob/fd2ef9c5ee891fdbc787e475359a48b3f6b0e094/controller/volume_controller.go#L3671-L3680
1. On the other hand, when user detach a volume, Longhorn will mark unscheduled replica as `failed` as part of `closeVolumeDependentResources` logic. See ref code: https://github.com/longhorn/longhorn-manager/blob/fd2ef9c5ee891fdbc787e475359a48b3f6b0e094/controller/volume_controller.go#L1988-L1996
1. Now `replica-3` has been marked as failed and has `replica.Spec.RebuildRetryCount = scheduler.FailedReplicaMaxRetryCount`, the replica clean up logic kick in and cleanup this replica. See ref code https://github.com/longhorn/longhorn-manager/blob/fd2ef9c5ee891fdbc787e475359a48b3f6b0e094/controller/volume_controller.go#L4872-L4876
This is the log:
    `
    [longhorn-manager-m2k9n longhorn-manager] time="2025-01-28T07:58:39Z" level=warning msg="Replica testvol-r-0c087bc9 failed to rebuild too many times" func="controller.(*VolumeController).shouldCleanUpFailedReplica" file="volume_controller.go:4880" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=phan-v713-pool2-5nr9q-78qnr owner=phan-v713-pool2-5nr9q-78qnr replica=testvol-r-0c087bc9 state=detaching volume=testvol
    `

cc @ChanYiLin @derekbit @innobead @shuo-wu @c3y1huang 

## Proposal solution 
TODO

**Comment by ChanYiLin** [comment]
Verified the **idea2** by setting the volume to ready for workload if all replicas are scheduled even if the `scheduledCondition` is false.

https://github.com/longhorn/longhorn-manager/blob/master/api/model.go#L1571
```
	if (v.Spec.NodeID == "" && v.Status.State != longhorn.VolumeStateDetached) ||
	     - (v.Status.State == longhorn.VolumeStateDetached && scheduledCondition.Status != longhorn.ConditionStatusTrue) ||
	     + (v.Status.State == longhorn.VolumeStateDetached && scheduledCondition.Status != longhorn.ConditionStatusTrue && !allReplicaScheduled(replicas)) ||
		v.Status.Robustness == longhorn.VolumeRobustnessFaulted ||
		v.Status.RestoreRequired ||
		(isCloningDesired && !isCloningCompleted) {
		ready = false
	}
```

cc @PhanLe1010 @derekbit

**Comment by derekbit** [comment]
@ChanYiLin Let's move the ticket and its backports to `Ready For Testing`. Thank you.

**Comment by roger-ryao** [comment]
**Verified on v1.9.0-dev-20250209 20250210**

- longhorn `v1.9.0-dev-20250209` https://github.com/longhorn/longhorn/commit/397b1381108b669e7a1d9bb679f4e40beee93973
- longhorn-manager `v1.9.0-dev-20250209` https://github.com/longhorn/longhorn-manager/commit/47ca86751f8865eeee19b76b63defa73dc008754

**The test steps**
https://github.com/longhorn/longhorn/issues/9938#issuecomment-2632574324

**Result Passed**
